debugtools/DDR_VM/src/com/ibm/j9ddr/tools/ddrinteractive/CacheMemorySource.java

```diff
@@ -0,0 +1,166 @@
+ * Secondary Licenses when the conditions for such availability set                                                                                                          
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU                                                                                                            
+ * General Public License, version 2 with the GNU Classpath                                                                                                                  
+ * Exception [1] and GNU General Public License, version 2 with the                                                                                                          
+ * OpenJDK Assembly Exception [2].                                                                                                                                           
+ *                                                                                                                                                                           
+ * [1] https://www.gnu.org/software/classpath/license.html                                                                                                                   
+ * [2] http://openjdk.java.net/legal/assembly-exception.html                                                                                                                 
+ *                                                                                                                                                                           
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception                                       
+ *******************************************************************************/
+package com.ibm.j9ddr.tools.ddrinteractive;
+
+import java.nio.ByteOrder;
+import java.nio.ByteBuffer;
+import java.nio.Buffer;
+import java.lang.reflect.Field;
+import com.ibm.j9ddr.corereaders.memory.MemoryFault;
+import com.ibm.j9ddr.corereaders.Platform;
+import com.ibm.j9ddr.corereaders.memory.AbstractMemory;
+import com.ibm.j9ddr.corereaders.memory.IMemorySource;
+import com.ibm.j9ddr.corereaders.memory.IMemoryRange;
+import com.ibm.j9ddr.corereaders.memory.ProtectedMemoryRange;
+import java.io.IOException;
+import java.nio.BufferUnderflowException;
+import sun.nio.ch.DirectBuffer;
+
+public class CacheMemorySource extends ProtectedMemoryRange implements IMemorySource
+        {
+
+	    private ByteBuffer bb;
+	    protected final int asid;
+	    protected final long baseAddress;
+	    protected final long size;	    
+
+	    public CacheMemorySource(long address,
+				int size)
+	    {
+
+		super(address,(long)size);
```

Comment:
I suggest the caller should supply the `ByteBuffer` and its base address. The buffer's position and limit should define the extent of this source. That would remove the need to make assumptions about the implementation of direct buffers and avoid the need to use reflection.

Reply:
sure, I think this makes sense. Related question: do you think we need a mechanism to guarantee the buffer is read-only or is it fine if its not, because we do not expose setter-ish methods anyways?

Reply:
If I understand your question, I think you're really asking about whether we should expect the caller to provide a buffer that doesn't change (independent of whether `isReadOnly()` returns `true`). I think a note in the javadoc that says the behavior is underined if that assumption is violated would be sufficient.